OSDOCS6774: Created new section to update support router to load secrets (#76392)
openshift/openshift-docs: modules/nw-ingress-route-secret-load-external-cert.adoc

// Module included in the following assemblies:
//
// * ingress/routes.adoc

:_mod-docs-content-type: PROCEDURE
[id="nnw-ingress-route-secret-load-external-cert_{context}"]
= Route Secret load For External Certificate

Currently, users of OpenShift cannot easily integrate OpenShift Routes with third-party certificate management solutions like cert-manager. The existing design of the Routes API mandates that certificate data be embedded within the Route object, rather than referenced. This creates challenges for seamless integration and automated lifecycle management.

> 

Reviewer: Do we need to talk about the previous challenge here? I'm not sure about the standard practice.

Author: Will be deleted.

> This enhancement enables a seamless integration between OpenShift Routes and third-party certificate management solutions such as cert-manager. By extending the Route API, users can reference externally managed certificates via secrets, eliminating the need for manual certificate management and reducing errors. The enhancement facilitates a smoother rollout of certificate updates, ensuring the OpenShift router serves renewed certificates promptly.

Reviewer: Which enhancement are we referring to here? Is it necessary for the end-user to know about the `o/enhancement` proposal?

// Module included in the following assemblies:
//
// * ingress/routes.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-ingress-route-secret-load-external-cert_{context}"]

> = Creating route with externally managed certificate

Reviewer 2: Should this be `Creating a route...`

:FeatureName: Securing route with external certificates in TLS secrets
include::snippets/technology-preview.adoc[]

> You can configure OpenShift routes with third-party certificate management solutions by using the `.spec.tls.externalCertificate` field of the route API. You can reference externally managed TLS certificates via secrets, eliminating the need for manual certificate management. Using the externally managed certificate reduces errors and provides faster updates. 

Reviewer 2: ```suggestion
You can configure {product-title} routes with third-party certificate management solutions by using the `.spec.tls.externalCertificate` field of the route API. You can reference externally managed TLS certificates via secrets, eliminating the need for manual certificate management. Using the externally managed certificate reduces errors and provides faster updates. 
```

[NOTE]
====
This feature applies to both edge routes and re-encrypt routes.
====

.Prerequisites

* You must enable the `RouteExternalCertificate` feature gate.

> * You must have the `create` and `update` permissions on the `routes/custom-host` sub-resource.

Reviewer 2: Does this just mean you need Admin priveleges?

> * You must have a secret contains a valid certificate/key pair in PEM-encoded format of type `kubernetes.io/tls`, which includes both `tls.key` and `tls.crt` keys.

Reviewer: ```suggestion
* You must have a secret containing a valid certificate/key pair in PEM-encoded format of type `kubernetes.io/tls`, which includes both `tls.key` and `tls.crt` keys.
```

* You must place the referenced secret in the same namespace as the route you want to secure.

.Procedure

> . Create a `role` in the same namespace as the secret to allow the router service account read access.

Reviewer 2: ```suggestion
. Create a `role` in the same namespace as the secret to allow the router service account read access by running the following command:
```

Reviewer 2: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#:~:text=If%20a%20step%20in%20a%20procedure%20is%20to%20run%20a%20command%2C%20make%20sure%20that%20the%20step%20text%20includes%20an%20explicit%20instruction%20to%20%22run%22%20or%20%22enter%22%20the%20command.%20In%20most%20cases%2C%20use%20one%20of%20the%20following%20patterns%20to%20introduce%20the%20code%20block%3A

+
[source,terminal]
----
$ oc create role secret-reader --verb=get,list,watch --resource=secrets --resource-name=<secret-name> \ <1> 
--namespace=<current-namespace> <2>
----

> <1> Replace <secret-name> with the actual name of your secret.

Reviewer 2: I would do this section using where: . [Here](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#:~:text=of%20requested%20CPU.-,If%20you%20must%20provide%20additional%20information%20on%20what%20a%20line%20of,if%20the%20same%20note%20applies%20to%20multiple%20lines%20of%20the%20codeblock.,-Be%20sure%20to) is a link to the contributing guidelines.

> <1> Replace <secret-name> with the actual name of your secret.

Reviewer 2: ```suggestion
+
where: <secret-name> is the name you have chosen for your secret.
```

<2> Replace <current-namespace> with the namespace where both your secret and route reside.

> . Create a `rolebinding` in the same namespace as the secret and bind the router service account to the newly created role.

Reviewer 2: Same comment here as the last step that a user needed to run a command.

+
[source,terminal]
----
$ oc create rolebinding secret-reader-binding --role=secret-reader --serviceaccount=openshift-ingress:router --namespace=<current-namespace> <1>
----

> <1> Replace <current-namespace> with the namespace where both your secret and route reside.

Reviewer 2: I would do  the `where:` option here as well.

> . Create a `route` specifying the `.spec.tls.externalCertificate` field to reference the secret containing your certificate:

Reviewer 2: I would recommend changing this step so that it is `. Create a YAML file that defines the route and specifies the secret containing your certificate using the following example`. Or something along those lines. I would have them create the YAML file because I'm guessing there might be a missing step where the file or YAML needs to be applied too. `oc apply -f <yaml_name>`.  I'm not an engineer though so I would check with devs on this.

Reviewer: That sounds perfect to me. Users need to apply the yaml after defining it. :+1:

Author: @chiragkyal Do we need to provide any specific instruction to do the apply yaml step?

Reviewer: Yeah something like

- Create a route resource by running the following command:
`oc apply -f <file.yaml>`
where `<file.yaml>` is your YAML file created above.

Feel free to change as per guidelines.

+
.YAML Definition of the Secure Route
[source,yaml]
----
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  name: myedge
  namespace: test
spec:
  host: myedge-test.apps.example.com
  tls:
    externalCertificate:
      name: <secret-name> <1>
    termination: edge
    [...]
[...]
----
<1> Replace <secret-name> with the actual name of your secret.

If the secret exists and has a certificate/key pair, the router will serve the generated certificate if all prerequisites are met.

[NOTE]
====

> If `.spec.tls.externalCertificate` is not provided, the router will use default generated certificates. You cannot provide both `.spec.tls.certificate`/`.spec.tls.key` and `.spec.tls.externalCertificate`.

Reviewer 2: ```suggestion
If `.spec.tls.externalCertificate` is not provided, the router will use default generated certificates. You cannot provide both the `.spec.tls.certificate` field  and the `.spec.tls.key` field if you are providing the `.spec.tls.externalCertificate` field.
```

Reviewer 2: Or you could rephrase to: You can only provide the `.spec.tls.certificate` field or the `.spec.tls.key` field if you are using the `.spec.tls.externalCertificate` field in your route configuration.

====

.Troubleshooting

> * If problems persist, check the OpenShift router pod logs for errors. For more information, refer to “investigating pod issues” for further information.

Reviewer 2: I would move this to additional resources and xref to it because the navigation can't be embedded in a module and having it makes for a better user experience, imo.

Author: Should I add additional resources in the module, following the [additional resources for modules guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#additional-resources-sections)? I already have added the xref in the additional resources in the assembly.

// Module included in the following assemblies:
//
// * networking/routes/secured-routes.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-ingress-route-secret-load-external-cert_{context}"]
= Creating a route with externally managed certificate

:FeatureName: Securing route with external certificates in TLS secrets
include::snippets/technology-preview.adoc[]

> You can configure {product-title} routes with third-party certificate management solutions by using the `.spec.tls.externalCertificate` field of the route API. You can reference externally managed TLS certificates via secrets, eliminating the need for manual certificate management. Using the externally managed certificate reduces errors and provides faster updates. 

Reviewer: I would agree with @lunarwhite to reword this a bit. https://github.com/openshift/openshift-docs/pull/76392#issuecomment-2163492232

* You must have the `create` and `update` permissions on the `routes/custom-host` sub-resource.
* You must have a secret containing a valid certificate/key pair in PEM-encoded format of type `kubernetes.io/tls`, which includes both `tls.key` and `tls.crt` keys.
* You must place the referenced secret in the same namespace as the route you want to secure.

.Procedure

. Create a `role` in the same namespace as the secret to allow the router service account read access by running the following command:
+
[source,terminal]
----
$ oc create role secret-reader --verb=get,list,watch --resource=secrets --resource-name=<secret-name> --namespace=<current-namespace> 
----
+
where: <secret-name> is the name you have chosen for your secret and <current-namespace> is the namespace where both your secret and route reside.

. Create a `rolebinding` in the same namespace as the secret and bind the router service account to the newly created role by running the following command:
+
[source,terminal]
----
$ oc create rolebinding secret-reader-binding --role=secret-reader --serviceaccount=openshift-ingress:router --namespace=<current-namespace>
----
+ 

> where: <secret-name> is the name you have chosen for your secret.

Reviewer: We don't have `<secret-name>` for the above command, it should be `<current-namespace>`

> . Create a `route` specifying the `.spec.tls.externalCertificate` field to reference the secret containing your certificate by running the following command:

Reviewer: > by running the following command:

I think users will need to first create the yaml, then run the `oc apply` command. I hope it will be changed once https://github.com/openshift/openshift-docs/pull/76392#discussion_r1636967481 suggestion is incorporated.

+
.YAML Definition of the Secure Route
[source,yaml]
----
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  name: myedge
  namespace: test
spec:
  host: myedge-test.apps.example.com
  tls:
    externalCertificate:
      name: <secret-name> 
    termination: edge
    [...]
[...]
----
+
where: <secret-name> is the name you have chosen for your secret.

If the secret exists and has a certificate/key pair, the router will serve the generated certificate if all prerequisites are met.

[NOTE]
====

> If `.spec.tls.externalCertificate` is not provided, the router will use default generated certificates. You can only provide the .spec.tls.certificate field or the .spec.tls.key field if you are using the .spec.tls.externalCertificate field in your route configuration.

Reviewer: > You can only provide the .spec.tls.certificate field or the .spec.tls.key field if you are using the .spec.tls.externalCertificate field in your route configuration.

Now, it sounds incorrect to me. 

The validation here is that: Users must neither provide `.spec.tls.certificate` field nor `.spec.tls.key` field if they are using `.spec.tls.externalCertificate` field. Feel free reword it so that it conveys the right information.

// Module included in the following assemblies:
//
// * ingress/routes.adoc

:_mod-docs-content-type: CONCEPT
[id="nnw-ingress-route-secret-load-external-cert_{context}"]

> = Route Secret load For External Certificate

Author: Sounds good to me

> You can now seamlessly integrate OpenShift Routes with third-party certificate management solutions. The extended Route API enables you to reference externally managed TLS certificates via secrets, eliminating the need for manual certificate management and reducing potential errors. This feature ensures a smoother rollout of certificate updates, enabling the OpenShift router to serve renewed certificates promptly.

Reviewer: I'm not aware of the docs standard for words like: `Routes`, `Route`, `API` : do they start with capital letters?

[id="nw-ingress-route-secret-load-external-cert_{context}"]
= Creating route with externally managed certificate

> You can configure OpenShift routes with third-party certificate management solutions by using the .spec.tls.externalCertificate field of the route API. You can reference externally managed TLS certificates via secrets, eliminating the need for manual certificate management. Using the externally managed certificate reduces errors and provides faster updates.

Reviewer: Code formatting 

> .spec.tls.externalCertificate

`.spec.tls.externalCertificate`
